rpcs3/Emu/Cell/PPUModule.cpp

```diff
@@ -694,6 +694,91 @@ static auto ppu_load_imports(std::vector<ppu_reloc>& relocs, ppu_linkage_info* l
 	return result;
 }
 
+static void ppu_check_patch_spu_images(const ppu_segment& seg)
+{
+	const std::string_view seg_view{vm::get_super_ptr<char>(seg.addr), seg.size};
+
+	for (std::size_t i = seg_view.find("\177ELF"); i < seg.size; i = seg_view.find("\177ELF", i + 4))
+	{
+		const auto elf_header = vm::get_super_ptr<u8>(seg.addr + i);
+
+		// Try to load SPU image
+		const spu_exec_object obj(fs::file(elf_header, seg.size - i));
+
+		if (obj != elf_error::ok)
+		{
+			// This address does not have an SPU elf
+			continue;
+		}
+
+		// Segment info dump
+		std::string dump;
+
+		std::size_t applied = 0;
+
+		// Executable hash
+		sha1_context sha2;
+		sha1_starts(&sha2);
+		u8 sha1_hash[20];
+
+		for (const auto& prog : obj.progs)
+		{
+			// Only hash the data, we are not loading it
+			sha1_update(&sha2, reinterpret_cast<const uchar*>(&prog.p_vaddr), sizeof(prog.p_vaddr));
+			sha1_update(&sha2, reinterpret_cast<const uchar*>(&prog.p_memsz), sizeof(prog.p_memsz));
+			sha1_update(&sha2, reinterpret_cast<const uchar*>(&prog.p_filesz), sizeof(prog.p_filesz));
+
+			fmt::append(dump, "\n\tSegment: p_type=0x%x, p_vaddr=0x%llx, p_filesz=0x%llx, p_memsz=0x%llx, p_offset=0x%llx", prog.p_type, prog.p_vaddr, prog.p_filesz, prog.p_memsz, prog.p_offset);
+
+			if (prog.p_type == 0x1u /* LOAD */ && prog.p_filesz > 0u)
+			{
+				sha1_update(&sha2, (elf_header + prog.p_offset), prog.p_filesz);
+			}
+
+			else if (prog.p_type == 0x4u /* NOTE */ && prog.p_filesz > 0u)
+			{
+				sha1_update(&sha2, (elf_header + prog.p_offset), prog.p_filesz);
+
+				// We assume that the string SPUNAME exists 0x14 bytes into the NOTE segment
+				const auto spu_name = reinterpret_cast<const char*>(elf_header + prog.p_offset + 0x14);
+				fmt::append(dump, "\n\tSPUNAME: '%s'", spu_name);
+			}
+		}
+
+		sha1_finish(&sha2, sha1_hash);
+
+		// Format patch name
+		std::string hash("SPU-0000000000000000000000000000000000000000");
+		for (u32 i = 0; i < sizeof(sha1_hash); i++)
+		{
+			constexpr auto pal = "0123456789abcdef";
+			hash[4 + i * 2] = pal[sha1_hash[i] >> 4];
+			hash[5 + i * 2] = pal[sha1_hash[i] & 15];
+		}
+
+		if (g_cfg.core.spu_debug)
+		{
+			fs::file dump_file(fs::get_cache_dir() + "/spu_progs/" + hash, fs::rewrite);
```

Comment:
perhaps, SPUNAME could be appended to hash somehow.

Reply:
Do we really want multiple files with the same data in such case? Theres also a case where there are multiple NOTE sections...
Im also not modifiying current hash for this as it breaks patches.

Reply:
Actually line 740 already does this.

Reply:
You misunderstood, I meant appending it to filename while keeping the hash. Sorry for the confusion.